### PR TITLE
Fixes for Nuvoton (GPIO)

### PR DESCRIPTION
**Pull request by dgarske**
* Refactor GPIO support (single gpio_config).
* Improved the Linux HAL IO layer try again logic.
* Improve the printing of the capabilities vendor firmware number.
* Spelling fixes.

### CODE REVIEW

File: examples/gpio/gpio_config.c [hunk 1]
```diff
@@ -227,15 +249,179 @@ int TPM2_GPIO_Config_Example(void* userCtx, int argc, char *argv[])
...
+    /* Configure NV Index for access to this GPIO */
+    XMEMSET(&nv, 0, sizeof(nv));
+    XMEMSET(&parent, 0, sizeof(parent));
+    /* Initial NV attributes */
+    parent.hndl = TPM_RH_PLATFORM;
+    rc = wolfTPM2_GetNvAttributesTemplate(parent.hndl, &nvAttributes);
+    /* Add NV attributes required by Nuvoton specification */
+    nvAttributes |= (TPMA_NV_PLATFORMCREATE | TPMA_NV_POLICY_DELETE);
+    nvAttributes |= (TPM_NT_ORDINARY & TPMA_NV_TPM_NT);
+    if (rc != 0) {
```
Comment (elms) [inline examples/gpio/gpio_config.c:399]:
```suggestion
    if (rc != TPM_RC_SUCCESS) {
```

File: examples/gpio/gpio_config.c [hunk 2]
```diff
@@ -227,15 +249,179 @@ int TPM2_GPIO_Config_Example(void* userCtx, int argc, char *argv[])
...
+        goto exit;
+    }
+#ifdef DEBUG_WOLFTPM
+    printf("nvAttributes = 0x%8.8X\n", nvAttributes);
+#endif
+
+    /* Define NV Index for GPIO */
+    rc = wolfTPM2_NVCreateAuth(&dev, &parent, &nv, nvIndex, nvAttributes,
+                               sizeof(BYTE), (byte*)gNvAuth, sizeof(gNvAuth)-1);
+    if (rc != 0 && rc != TPM_RC_NV_DEFINED) {
```
Comment (elms) [inline examples/gpio/gpio_config.c:410]:
```suggestion
    if (rc != TPM_RC_SUCCESS && rc != TPM_RC_NV_DEFINED) {
```

File: examples/gpio/gpio_config.c [hunk 3]
```diff
@@ -227,15 +249,179 @@ int TPM2_GPIO_Config_Example(void* userCtx, int argc, char *argv[])
...
+        rc = wolfTPM2_StartSession(&dev, &tpmSessionIndex, NULL, NULL,
+                                   TPM_SE_POLICY, TPM_ALG_NULL);
+        if (rc == TPM_RC_SUCCESS) {
+            printf("index ok\n");
+        }
+
+        rc = wolfTPM2_StartSession(&dev, &tpmSessionPlatform, NULL, NULL,
+                                   TPM_SE_POLICY, TPM_ALG_NULL);
+
+        if (rc == TPM_RC_SUCCESS) {
```
Comment (elms) [inline examples/gpio/gpio_config.c:282]: Is there a reason to not `goto exit` here on error? Same on previous check for startsession?

File: src/tpm2.c [hunk 4]
```diff
@@ -5284,7 +5286,9 @@ int TPM2_GPIO_Config(GpioConfig_In* in)
     }
     return rc;
 }
-#elif defined(WOLFTPM_NUVOTON)
+#endif /* WOLFTPM_ST33 */
+
+#ifdef WOLFTPM_NUVOTON
```
Comment (elms) [inline src/tpm2.c:5291]: Why change to exclusive logic? It seems like only one can be compile in at a time.

**Review by elms — CHANGES_REQUESTED** [review]
Looks good, but I would like to see either configure step fail or the build with both ST and nuvoton to complete.

**Comment by dgarske** [comment]
@elms , fixes pushed. Thanks for the excellent review.

**Review by elms — APPROVED** [review]

**Pull request merged**